<launch>
  <!-- Robot 1 Controller -->
  <group ns="arm1">
    <!-- Load joint controller configurations from YAML file to parameter server -->
    <rosparam file="$(find arm_control)/config/arm_controller.yaml" command="load"/>

    <!-- Load robot description for this namespace -->
    <param name="robot_description" command="$(find xacro)/xacro --inorder '$(find arm_description)/urdf/lbr_iiwa_14_r820.xacro' origin_xyz:='0 0 0'" />
    <!-- Also load it into the controller's private namespace so it can find it -->
    <!-- NOTE: We use origin 0 0 0 for the controller URDF so all arms share the same model. 
         The actual position is handled by the base_pos parameters. -->
    <param name="arm_controller/robot_description" command="$(find xacro)/xacro --inorder '$(find arm_description)/urdf/lbr_iiwa_14_r820.xacro' origin_xyz:='0 0 0'" />
    
    <!-- Controller Spawner -->
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
          output="screen" args="arm_controller joint_state_controller"/>
          
    <!-- Robot State Publisher -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
          respawn="false" output="screen">
      <param name="tf_prefix" value="arm1" />
    </node>

    <!-- Trajectory Publisher for Arm 1 -->
    <node name="traj_publisher" pkg="arm_control" type="traj_publisher_node" output="screen">
      <remap from="/command_pose" to="/arm1/arm_control/command_pose"/>
      <remap from="/command_twist" to="/arm1/arm_control/command_twist"/>
      <!-- Customizing parameters to match arm setup -->
      <param name="center_x" value="0.6"/>
      <param name="center_y" value="0.0"/>
      <param name="center_z" value="0.4"/>
      <param name="radius" value="0.1"/>
      <param name="freq" value="0.5"/>
    </node>
  </group>

  <!-- Robot 2 Controller -->
  <group ns="arm2">
    <rosparam file="$(find arm_control)/config/arm_controller.yaml" command="load"/>
    <param name="arm_controller/recompile_libraries" value="false"/>

    <param name="robot_description" command="$(find xacro)/xacro --inorder '$(find arm_description)/urdf/lbr_iiwa_14_r820.xacro' origin_xyz:='1 0 0'" />
    <param name="arm_controller/robot_description" command="$(find xacro)/xacro --inorder '$(find arm_description)/urdf/lbr_iiwa_14_r820.xacro' origin_xyz:='0 0 0'" />
    
    <!-- Delay start by 20 seconds -->
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
          output="screen" args="arm_controller joint_state_controller"
          launch-prefix="bash -c 'sleep 20; $0 $@' "/>
          
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
          respawn="false" output="screen">
      <param name="tf_prefix" value="arm2" />
    </node>
  </group>

  <!-- Robot 3 Controller -->
  <group ns="arm3">
    <rosparam file="$(find arm_control)/config/arm_controller.yaml" command="load"/>
    <param name="arm_controller/recompile_libraries" value="false"/>

    <param name="robot_description" command="$(find xacro)/xacro --inorder '$(find arm_description)/urdf/lbr_iiwa_14_r820.xacro' origin_xyz:='0 1 0'" />
    <param name="arm_controller/robot_description" command="$(find xacro)/xacro --inorder '$(find arm_description)/urdf/lbr_iiwa_14_r820.xacro' origin_xyz:='0 0 0'" />
    
    <!-- Delay start by 40 seconds -->
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
          output="screen" args="arm_controller joint_state_controller"
          launch-prefix="bash -c 'sleep 40; $0 $@' "/>
          
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
          respawn="false" output="screen">
      <param name="tf_prefix" value="arm3" />
    </node>
  </group>

</launch>